### PR TITLE
Add KV replicate feature when kv_head<tp_size

### PR DESCRIPTION
**Pull request by kaixuanliu**
[CPU CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47811) [GPU run-slow](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47811)

### CODE REVIEW

**Comment by kaixuanliu** [comment]
@Cyrilvallez @3outeille This PR tries to support 1 scenario: when the model need to be loaded with tp mode, but the tp_size is larger than kv_head, we need to do KV replicate like the feature in VLLM. I have verified the correctness of this PR using `bigcode/starcoder2-3b` model w/ kv_head=2 and tp_size=4. pls help review, thx!

**Comment by 3outeille** [comment]
Hey super cool ! Just a note that we are undergoing a refactor of our Tensor Parallel cf https://github.com/huggingface/transformers/pull/47579.  If that's not too much, is it possible to rebase your work on top of the new branch ?

**Comment by kaixuanliu** [comment]
Here is the example code:
```
import os

import torch
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh

from transformers import AutoModelForCausalLM, AutoTokenizer
from transformers.distributed.configuration_utils import DistributedConfig


MODEL_ID = "bigcode/starcoder2-3b"
PROMPT = "Explain tensor parallelism in one concise sentence."
MAX_NEW_TOKENS = 128


def main():
    local_rank = int(os.environ["LOCAL_RANK"])
    torch.cuda.set_device(local_rank)
    device = torch.device(f"cuda:{local_rank}")
    dist.init_process_group(backend="nccl")

    try:
        world_size = dist.get_world_size()
        tp_mesh = init_device_mesh("cuda", (world_size,), mesh_dim_names=("tp",))

        tokenizer = AutoTokenizer.from_pretrained(MODEL_ID)
        model = AutoModelForCausalLM.from_pretrained(
            MODEL_ID,
            device_mesh=tp_mesh,
            distributed_config=DistributedConfig(tp_size=world_size),
            dtype=torch.bfloat16,
            attn_implementation="sdpa",
        ).eval()
    
        inputs = tokenizer(PROMPT, return_tensors="pt").to(device)
        output_ids = model.generate(
            **inputs,
            max_new_tokens=MAX_NEW_TOKENS,
            do_sample=False,
            use_cache=True,
            cache_implementation="static",
        )

        # TP ranks must produce the same complete sequence. This catches missing all-reduces as well as incorrect
        # KV head replication/repeat_kv handling during cached decoding.
        gathered_output_ids = [torch.empty_like(output_ids) for _ in range(world_size)]
        dist.all_gather(gathered_output_ids, output_ids)
        if not all(torch.equal(output_ids, other) for other in gathered_output_ids[1:]):
            raise AssertionError("TP ranks generated different token sequences.")

        if dist.get_rank() == 0:
            completion = tokenizer.decode(output_ids[0, inputs.input_ids.shape[-1] :], skip_special_tokens=True)
            print(f"PASS: {MODEL_ID} generated consistently with TP={world_size} and KV head replication.")
            print(f"Completion: {completion}")
    finally:
        dist.destroy_process_group()


if __name__ == "__main__":
    main()
```

**Comment by kaixuanliu** [comment]
> Hey super cool ! Just a note that we are undergoing a refactor of our Tensor Parallel cf #47579. If that's not too much, is it possible to rebase your work on top of the new branch ?

Sure, will do the rebase work after https://github.com/huggingface/transformers/pull/47579 is merged.

**Comment by 3outeille** [comment]
Hey @kaixuanliu, PR has been merged ! Feel free to rebase, i'll review it

**Comment by kaixuanliu** [comment]
@3outeille ,rebase work is done, pls help review, thx!

File: src/transformers/distributed/tensor_parallel.py [hunk 1]
```diff
@@ -794,10 +891,71 @@ def _validate_tp_plan_styles(tp_plan: dict[str, str] | None) -> None:
...
+        replicated_modules.append(module_name)
+
+    if not replicated_modules:
+        return tp_plan
+
+    tp_plan = {
+        key: ("colwise_replicate_kv" if style == "colwise" and key.endswith(("k_proj", "v_proj")) else style)
+        for key, style in tp_plan.items()
+    }
+    logger.warning_once(
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: only one rank should log this

File: src/transformers/distributed/tensor_parallel.py [hunk 2]
```diff
@@ -241,6 +241,102 @@ def transform_output_post_forward(self, module, output, mesh):
...
+            f"num_key_value_heads={num_key_value_heads}: when there are fewer KV heads than ranks, "
+            f"`num_key_value_heads` must divide `tp_size` so that KV heads can be replicated."
+        )
+    return tp_size // num_key_value_heads
+
+
+_KV_REPLICATION_GROUPS: dict[tuple[int, int], object] = {}
+
+
+def _get_kv_replication_group(mesh, n_rep: int):
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: move it under ReplicateKVHeadsParallel as it only use in it

File: src/transformers/distributed/tensor_parallel.py [hunk 3]
```diff
@@ -241,6 +241,102 @@ def transform_output_post_forward(self, module, output, mesh):
...
+    ``dist.new_group`` is collective, so every rank walks through all the groups in the same order and keeps the
+    one it belongs to. The result is cached because every attention layer asks for the same group.
+    """
+    key = (id(mesh), n_rep)
+    if key not in _KV_REPLICATION_GROUPS:
+        global_ranks = mesh.mesh.flatten().tolist()
+        local_rank = mesh.get_local_rank()
+        group = None
+        for start in range(0, len(global_ranks), n_rep):
+            candidate = dist.new_group(ranks=global_ranks[start : start + n_rep])
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: does that still holds for 2D mesh ?

File: src/transformers/distributed/tensor_parallel.py [hunk 4]
```diff
@@ -241,6 +241,102 @@ def transform_output_post_forward(self, module, output, mesh):
...
+_KV_REPLICATION_GROUPS: dict[tuple[int, int], object] = {}
+
+
+def _get_kv_replication_group(mesh, n_rep: int):
+    """Process group holding the `n_rep` ranks that own the same (replicated) KV head.
+
+    ``dist.new_group`` is collective, so every rank walks through all the groups in the same order and keeps the
+    one it belongs to. The result is cached because every attention layer asks for the same group.
+    """
+    key = (id(mesh), n_rep)
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: double check because iirc `id(mesh)` as a key doesnt work. Had issue back in the day

```
import torch.distributed as dist
from torch.distributed.device_mesh import init_device_mesh

dist.init_process_group("gloo")
mesh = init_device_mesh("cpu", (2,), mesh_dim_names=["tp"])
a, b = mesh["tp"], mesh["tp"]

print(id(a) is id(b)) # False
```

File: src/transformers/distributed/tensor_parallel.py [hunk 5]
```diff
@@ -241,6 +241,102 @@ def transform_output_post_forward(self, module, output, mesh):
...
+        if n_rep > 1:
+            group = _get_kv_replication_group(mesh, n_rep)
+
+            # A module hook rather than `param.register_hook`: params are replaced during weight loading, which
+            # happens after TP is applied, and would drop a param-level hook.
+            def _all_reduce_replicated_grads(mod, grad_input, grad_output):
+                for param in mod.parameters(recurse=False):
+                    if param.grad is not None:
+                        grad = param.grad
+                        dist.all_reduce(grad.to_local() if isinstance(grad, DTensor) else grad, group=group)
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: to double check it works, pick a model with `tp_size > num_kv_heads` (i.e: `"Qwen/Qwen2.5-VL-3B-Instruct"`) and add it to `TP_DISTRIBUTED_TEST_MODEL_TYPES`  in `test_tensor_parallel_mixin.py`    and test its forward + backward to see if everything pass

I think we can leave the model in the list so that we can catch KV_replication regression later

File: src/transformers/distributed/sharding_utils.py [hunk 6]
```diff
@@ -123,6 +126,8 @@ def shard_tensor(
                 sub_mesh = self._get_sub_mesh(mesh_dim)
                 rank, world_size = sub_mesh.get_local_rank(), sub_mesh.size()
                 dim_idx = self._normalize_param_dim(placement.dim)
+                if self.kv_replication > 1 and placement.is_shard():
+                    rank, world_size = rank // self.kv_replication, world_size // self.kv_replication
```
Comment (3outeille) [inline src/transformers/distributed/sharding_utils.py]: correct for 1D, but not for 2D mesh

File: src/transformers/distributed/tensor_parallel.py [hunk 7]
```diff
@@ -794,10 +891,71 @@ def _validate_tp_plan_styles(tp_plan: dict[str, str] | None) -> None:
...
+    )
+    return tp_plan
+
+
 def apply_tensor_parallelism(model, tp_mesh):
     """DTensor backend: shard params as placeholders and install TP forward hooks."""
 
     _validate_tp_plan_styles(model.tp_plan)
+    if model.tp_plan is not None:
+        model.tp_plan = _maybe_enable_kv_head_replication(model, model.tp_plan, tp_mesh.size())
```
Comment (3outeille) [inline src/transformers/distributed/tensor_parallel.py]: if `enable_expert_parallel=True`, it will swap out `model.tp_plan` to `ep_plan`. Double check if replication is still applied

File: src/transformers/distributed/tensor_parallel.py (hunk 1 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: Fixed

File: src/transformers/distributed/tensor_parallel.py (hunk 2 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: Fixed

File: src/transformers/distributed/tensor_parallel.py (hunk 3 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: pls help review again

File: src/transformers/distributed/tensor_parallel.py (hunk 4 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: Same as above

File: src/transformers/distributed/sharding_utils.py (hunk 6 above)
Comment (kaixuanliu) [reply src/transformers/distributed/sharding_utils.py]: Since `apply_tensor_parallelism` is always handed device_mesh["tp"] and TP/FSDP/PP are mutually exclusive, this is unreachable today: [L194-L206](https://github.com/huggingface/transformers/blob/main/src/transformers/distributed/mixin.py#L194-L206), I have added a guard in `__init__` func

File: src/transformers/distributed/tensor_parallel.py (hunk 7 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: Thx for advice!! Have updated.

File: src/transformers/distributed/tensor_parallel.py (hunk 5 above)
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py]: Added `test_tp_kv_head_replication` to `TensorParallelTesterMixin`: it forces `num_key_value_heads=1` on the tiny config so num_kv_heads < tp_size, asserts the layers actually went through `ReplicateKVHeadsParallel`, and then runs the full forward + backward (incl. per-parameter grad comparison). This runs for every model already in `TP_DISTRIBUTED_TEST_MODEL_TYPES` (qwen2/qwen3/qwen3_moe/...), so we get regression coverage without having to wire the TP mixin into a VLM test class (`Qwen2_5_VLModelTest` doesn't use `CausalLMModelTester`, so it would be skipped anyway).

File: src/transformers/distributed/mixin.py [hunk 8]
```diff
@@ -266,7 +266,12 @@ def gather_sharded_state_dict_for_save(
 
         if distributed_config.tp_size > 1:
             state_dict = gather_state_dict_for_save(
-                state_dict, self._tp_plan, self._device_mesh, distributed_config.tp_size
+                state_dict,
+                # The property, not `_tp_plan`: under expert parallelism the plan that was applied is `_ep_plan`.
```
Comment (ArthurZucker) [inline src/transformers/distributed/mixin.py]:
```suggestion
```

File: src/transformers/distributed/sharding_utils.py [hunk 9]
```diff
@@ -81,10 +81,21 @@ class DtensorShardOperation:
        we skip it. Later, `MergeModulelist` will stack the owned expert we kept to create the rank's local shard
     """
 
-    def __init__(self, param: DTensor):
+    def __init__(self, param: DTensor, kv_replication: int = 1):
```
Comment (ArthurZucker) [inline src/transformers/distributed/sharding_utils.py]: I really don't think it makes sense to put this attention, kv specific argument into a very general shading scheme.

File: src/transformers/distributed/tensor_parallel.py
Comment (ArthurZucker) [inline src/transformers/distributed/tensor_parallel.py:1]: this is highly bloated, and that makes me think it just does not follow the design.
We are probably missing ShardingOps, that you would want for that, otherwise we can just properly resolve, based on the dim, what to do with the weights, regardless of attention or not. The same happens for WP sharding:  see #48237

**Review by ArthurZucker — COMMENTED** [review]
Really not aligned as this IMO should be much much simpler to do

File: src/transformers/distributed/sharding_utils.py (hunk 9 above)
Comment (kaixuanliu) [reply src/transformers/distributed/sharding_utils.py]: Well, I agree. It is not a good design indeed..., have fixed it.

File: src/transformers/distributed/tensor_parallel.py
Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py:1]: Well the diff is a little big. But I don't think a purely dim-based rule can cover this case. Also take `bigcode/starcoder2-3b` as example, `num_key_value_heads=2`, `head_dim=128`, so `k_proj.out_features=256`. On `tp=4, 256 % 4 == 0` — a dim-based rule shards it happily, and attention then dies on k_proj(x).view(*input_shape, -1, self.head_dim) with 64 features per rank. Whenever kv_heads < tp_size you get less than one head per rank, so this always happens. It can't be fixed in attention either: rotate_half pairs dim i with i + head_dim/2, and qk/softmax/FA kernels all need a whole head. head_dim is a hard granularity floor, so replication is the only option.

Comment (kaixuanliu) [reply src/transformers/distributed/tensor_parallel.py:1]: @ArthurZucker I have refactored the code to align with your comments: the sharding machinery is based on head_dim.

**Review by VI-Arthur — CHANGES_REQUESTED** [review]
I'll work on something simpler, this is too much attention specific code when you want this to behave the same for all layers. Ty for raising tho, it is important

**Comment by ArthurZucker** [comment]
CF linked PR!
